### PR TITLE
Bugfix/nextcloud cmd dav fix

### PR DESCRIPTION
**Pull request by allexzander**
This should fix https://github.com/nextcloud/desktop/issues/3616

### CODE REVIEW

File: src/cmd/cmd.cpp [hunk 1]
```diff
@@ -442,6 +441,7 @@ int main(int argc, char **argv)
     }
 #endif
 
+    const auto UrlString = url.toString();
```
Comment (FlexW) [inline src/cmd/cmd.cpp:444]: urlString. Where is this used?

File: src/cmd/cmd.cpp [hunk 2]
```diff
@@ -350,6 +350,9 @@ int main(int argc, char **argv)
 
     QUrl url = QUrl::fromUserInput(options.target_url);
 
+    // take only the base URL (without the path)
+    url = url.toString().remove(url.path());
```
Comment (FlexW) [inline src/cmd/cmd.cpp:354]: maybe you could rename it to baseUrl

File: src/cmd/cmd.cpp [hunk 3]
```diff
@@ -388,24 +391,20 @@ int main(int argc, char **argv)
...
-    const QStringList splitted = url.path().split(account->davPath());
-    url.setPath(splitted.value(0));
 
     url.setScheme(url.scheme().replace("owncloud", "http"));
 
     QUrl credentialFreeUrl = url;
     credentialFreeUrl.setUserName(QString());
     credentialFreeUrl.setPassword(QString());
 
+    const auto urlWithoutTrailingSlash = options.target_url.endsWith('/') ? options.target_url.chopped(1) : options.target_url;
```
Comment (FlexW) [inline src/cmd/cmd.cpp:402]: QLatin1Char

File: src/cmd/cmd.cpp [hunk 4]
```diff
@@ -388,24 +391,20 @@ int main(int argc, char **argv)
...
+
+    const auto urlSplitted = urlWithoutTrailingSlash.split('/', QString::SplitBehavior::SkipEmptyParts);
```
Comment (FlexW) [inline src/cmd/cmd.cpp:404]: QLatin1Char

File: src/cmd/cmd.cpp [hunk 5]
```diff
@@ -388,24 +391,20 @@ int main(int argc, char **argv)
...
+
     // Remote folders typically start with a / and don't end with one
-    QString folder = "/" + splitted.value(1);
-    if (folder.endsWith("/") && folder != "/") {
-        folder.chop(1);
-    }
+    const auto folder = urlSplitted.size() > 1 ? QString("/" + urlSplitted.last()) : QLatin1String("/");
```
Comment (FlexW) [inline src/cmd/cmd.cpp:407]: QStringLiteral

File: src/libsync/account.cpp [hunk 6]
```diff
@@ -84,7 +84,9 @@ Account::~Account() = default;
 
 QString Account::davPath() const
 {
-    return QLatin1String("/remote.php/dav/files/") + davUser() + QLatin1Char('/');
+    const auto davPathBase = QLatin1String("/remote.php/dav/files/");
+    const auto davPathFinal = !davUser().isEmpty() ? QString(davPathBase + davUser()) + QLatin1Char('/') : QString(davPathBase);
```
Comment (FlexW) [inline src/libsync/account.cpp:88]: QString(davPathBase) is redundant or?

File: src/libsync/account.cpp [hunk 7]
```diff
@@ -84,7 +84,9 @@ Account::~Account() = default;
 
 QString Account::davPath() const
 {
-    return QLatin1String("/remote.php/dav/files/") + davUser() + QLatin1Char('/');
+    const auto davPathBase = QLatin1String("/remote.php/dav/files/");
```
Comment (FlexW) [inline src/libsync/account.cpp:87]: const QString davPathBase(...) would avoid one assignment.

File: src/cmd/cmd.cpp (hunk 1 above)
Comment (allexzander) [reply src/cmd/cmd.cpp:444]: @FlexW Sorry, debugging leftovers :)

File: src/libsync/account.cpp (hunk 6 above)
Comment (allexzander) [reply src/libsync/account.cpp:88]: @FlexW Conversion is here just to avoid a problem we had in past when returning `QStringBuilder` instead of `QString`, or, I did not get your comment ?

File: src/libsync/account.cpp [hunk 8]
```diff
@@ -84,7 +84,9 @@ Account::~Account() = default;
 
 QString Account::davPath() const
 {
-    return QLatin1String("/remote.php/dav/files/") + davUser() + QLatin1Char('/');
+    const QString davPathBase("/remote.php/dav/files/");
+    const QString davPathFinal = !davUser().isEmpty() ? davPathBase + davUser() + QLatin1Char('/') : davPathBase;
```
Comment (mgallien) [inline src/libsync/account.cpp:88]: this will generate wrong path in case `davUser()` is empty (OK I concede that the existing code will also fail but it has one less conditional)
why is this change needed ?

Comment (allexzander) [reply src/libsync/account.cpp:88]: @mgallien Well, this allows us to always have a proper format for the path. At the moment, we are getting `/remote.php/dav/files//` in case `davUser()`, which is incorrect. It has double `/`.

File: src/cmd/cmd.cpp [hunk 9]
```diff
@@ -339,7 +339,14 @@ int main(int argc, char **argv)
         qFatal("Could not initialize account!");
         return EXIT_FAILURE;
     }
-    // check if the webDAV path was added to the url and append if not.
+
+    if (options.target_url.contains("webdav")) {
+        qWarning("Webdav is detected, even though it is deprecated.");
```
Comment (allexzander) [inline src/cmd/cmd.cpp:344]: @mgallien As you requested.

File: src/cmd/cmd.cpp [hunk 10]
```diff
@@ -339,7 +339,14 @@ int main(int argc, char **argv)
...
+        std::cout << "Warning: Webdav is deprecated and is going to be removed soon. Please, use dav path for a host. Example:" << std::endl
+                  << "https://yourusername:youruserpassword@yourhost/remote.php/dav/files/remotefolderpath" << std::endl;
+    }
+
+    // check if the dav path was added to the url and append if not.
```
Comment (mgallien) [inline src/cmd/cmd.cpp:349]: What is the relation between this new comment and the code just after it ?
I am missing the point here

File: src/cmd/cmd.cpp [hunk 11]
```diff
@@ -348,16 +355,19 @@ int main(int argc, char **argv)
         options.target_url.append(account->davPath());
     }
 
-    QUrl url = QUrl::fromUserInput(options.target_url);
+    QUrl baseUrl = QUrl::fromUserInput(options.target_url);
+
+    // take only the base URL (without the path)
+    baseUrl = baseUrl.toString().remove(baseUrl.path());
```
Comment (mgallien) [inline src/cmd/cmd.cpp:361]: this will break if the user is deployed to a path other than the root path of the webserver
imagine something like `https://www.example.com/cloud/

File: src/cmd/cmd.cpp [hunk 12]
```diff
@@ -388,24 +398,20 @@ int main(int argc, char **argv)
         }
     }
 
-    // take the unmodified url to pass to csync_create()
-    QByteArray remUrl = options.target_url.toUtf8();
-
     // Find the folder and the original owncloud url
-    QStringList splitted = url.path().split("/" + account->davPath());
```
Comment (mgallien) [inline src/cmd/cmd.cpp:395]: why do you remove this instead of fixing it such that the value returned by davPath() is valid ?

File: src/cmd/cmd.cpp (hunk 9 above)
Comment (mgallien) [reply src/cmd/cmd.cpp:344]: Can you improve the message to not imply that webdav is deprectaed (webdav as the webdav protocol) but that the proper URL to use should include dav instead of webdav (as shown in the web UI) ?

File: src/libsync/account.cpp (hunk 8 above)
Comment (mgallien) [reply src/libsync/account.cpp:90]: I miss the advantage of returning an invalid URL that is the value of `davPathBase`
why do you want to return it since sync will not work with it ?

**Review by mgallien — CHANGES_REQUESTED** [review]

File: src/cmd/cmd.cpp (hunk 10 above)
Comment (allexzander) [reply src/cmd/cmd.cpp:349]: @mgallien No relation at all. It was there before and I don't need to remove it for my PR. It may seem to you that it was somehow modified, but, in fact, it is the wrong impression of Git diff. My comment was added before this line. So, I am now confused about, what concern you have and what I need to change?

Comment (mgallien) [reply src/cmd/cmd.cpp:355]: I will have a look again

File: src/cmd/cmd.cpp [hunk 13]
```diff
@@ -339,25 +344,29 @@ int main(int argc, char **argv)
...
         return EXIT_FAILURE;
     }
-    // check if the webDAV path was added to the url and append if not.
-    if (!options.target_url.endsWith("/")) {
-        options.target_url.append("/");
+
+    if (options.target_url.contains("/webdav", Qt::CaseInsensitive) || options.target_url.contains("/dav", Qt::CaseInsensitive)) {
+        qWarning("Dav or webdav in server URL.");
+        std::cout << "Error! Please specify only the base URL of your host with username and password. Example:" << std::endl
+                  << "https://username:password@http(s)://cloud.example.com" << std::endl;
```
Comment (mgallien) [inline src/cmd/cmd.cpp:351]: you have an extra `http(s)://` in the middle

File: src/cmd/cmd.cpp [hunk 14]
```diff
@@ -193,6 +194,7 @@ void help()
     std::cout << "  -h                     Sync hidden files, do not ignore them" << std::endl;
     std::cout << "  --version, -v          Display version and exit" << std::endl;
     std::cout << "  --logdebug             More verbose logging" << std::endl;
+    std::cout << "  --path                 Path to a folder on a remote server" << std::endl;
```
Comment (mgallien) [inline src/cmd/cmd.cpp:197]: I really like the direction this PR is taking !
I hope this change will make it easier to maintain `nextcloudcmd` in the future

**Review by mgallien — COMMENTED** [review]
a last small comment and from my point of view we are good to merge

**Review by mgallien — APPROVED** [review]

**Comment by mgallien** [comment]
please do not forget to clean history before merge

**Comment by allexzander** [comment]
> please do not forget to clean history before merge

Good point.

File: src/cmd/cmd.cpp [hunk 15]
```diff
@@ -339,25 +344,29 @@ int main(int argc, char **argv)
         qFatal("Could not initialize account!");
         return EXIT_FAILURE;
     }
-    // check if the webDAV path was added to the url and append if not.
-    if (!options.target_url.endsWith("/")) {
-        options.target_url.append("/");
+
+    if (options.target_url.contains("/webdav", Qt::CaseInsensitive) || options.target_url.contains("/dav", Qt::CaseInsensitive)) {
+        qWarning("Dav or webdav in server URL.");
+        std::cout << "Error! Please specify only the base URL of your host with username and password. Example:" << std::endl
```
Comment (FlexW) [inline src/cmd/cmd.cpp:350]: Should it not be `cerr` ?

File: src/libsync/account.h [hunk 16]
```diff
@@ -125,6 +125,8 @@ class OWNCLOUDSYNC_EXPORT Account : public QObject
      */
     QString davPath() const;
 
+    static QString davPathBase();
```
Comment (FlexW) [inline src/libsync/account.h:128]: would it make sense to make that one private or even put into the `.cpp` file? I don't think that needs to be part of the public api.

File: src/cmd/cmd.cpp [hunk 17]
```diff
@@ -442,7 +442,7 @@ int main(int argc, char **argv)
     }
 #endif
 
-    account->setUrl(url);
+    account->setUrl(hostUrl.toString().endsWith(QLatin1Char('/')) ? QUrl(hostUrl.toString().chopped(1)) : hostUrl);
```
Comment (FlexW) [inline src/cmd/cmd.cpp:445]: Is that really needed? I mean QUrl should not care if it ends with `/` or not.

File: src/cmd/cmd.cpp [hunk 18]
```diff
@@ -339,25 +344,29 @@ int main(int argc, char **argv)
...
 
-    if (!options.target_url.contains(account->davPath())) {
-        options.target_url.append(account->davPath());
+    // check if the dav path was added to the url and append if not.
+    if (!options.target_url.endsWith("/")) {
+        options.target_url.append("/");
     }
 
-    QUrl url = QUrl::fromUserInput(options.target_url);
+    QUrl hostUrl = QUrl::fromUserInput(options.target_url);
```
Comment (FlexW) [inline src/cmd/cmd.cpp:360]: Could be const?

File: src/libsync/account.h (hunk 16 above)
Comment (allexzander) [reply src/libsync/account.h:128]: @FlexW No. It is used in cmd.cpp :(

File: src/cmd/cmd.cpp (hunk 18 above)
Comment (allexzander) [reply src/cmd/cmd.cpp:360]: @FlexW Later, we are calling `setScheme` which is non-const method.

File: src/cmd/cmd.cpp (hunk 17 above)
Comment (allexzander) [reply src/cmd/cmd.cpp:445]: @FlexW For consistency. When running a GUI version, the URL set to an account is without the traling shash.

File: src/libsync/account.h (hunk 16 above)
Comment (allexzander) [reply src/libsync/account.h]: @FlexW Made it private.

File: src/cmd/cmd.cpp (hunk 17 above)
Comment (FlexW) [reply src/cmd/cmd.cpp:445]: But does it work without it? If it works without it I would prefer to drop it because it would improve readability. What do you think?

File: src/cmd/cmd.cpp (hunk 18 above)
Comment (FlexW) [reply src/cmd/cmd.cpp:360]: Sorry I overlooked it

File: src/cmd/cmd.cpp (hunk 17 above)
Comment (allexzander) [reply src/cmd/cmd.cpp:445]: @FlexW ok

**Review by FlexW — APPROVED** [review]

**Comment by mgallien** [comment]
because I forgot to ask the documentation update, please do not forget
